stdlib/stack_trace.mli

```diff
@@ -0,0 +1,141 @@
+      by the compiler, and [false] when it comes from any other context. *)
+
+  type location =
+    { filename: string;
+      line_number: int;
+      start_char: int;
+      end_char: int }
+  (** The type of location information found in stack traces. [start_char] and
+      [end_char] are positions relative to the beginning of the line. *)
+
+  val location: t -> location option
+  (** [location slot] returns the location information of the slot,
+      if available, and [None] otherwise.
+
+      Some possible reasons for failing to return a location are as follow:
+      - the slot corresponds to a compiler-inserted raise
+      - the slot corresponds to a part of the program that has not been
+      compiled with debug information ([-g]) *)
+
+  val format: int -> t -> string option
+  (** [format pos slot] returns the string representation of [slot] as
+      {!to_string} would format it, assuming it is the [pos]-th element of the
+      callstack: the [0]-th element is pretty-printed differently than the
+      others.
+
+      Whole-stack trace printing functions also skip some uninformative slots;
+      in that case, [format pos slot] returns [None]. *)
+
+  module Raw : sig
+    (** {1 Raw stack trace slots} *)
+
+    type t
+    (** This type allows direct access to raw stack trace slots, without any
+        conversion in an OCaml-usable data-structure. Being process-specific,
+        they must absolutely not be marshalled, and are unsafe to use for this
+        reason (marshalling them may not fail, but un-marshalling and using the
+        result will result in undefined behavior).
+
+        Elements of this type can still be compared and hashed: when two
+        elements are equal, then they represent the same source location (the
+        converse is not necessarily true in presence of inlining, for
+        example). *)
+
+    val get: stack_trace -> int -> t
+    (** [get st pos] returns the slot in position [pos] in the stack trace
+        [st]. *)
+
+    val next: t -> t option
+    (** [next slot] returns the next inlined slot, if any.
+
+        Sample code to iterate over all frames (inlined and non-inlined):
+        {[
+          (* Iterate over inlined frames *)
+          let rec iter_callstack_slot f slot =
+            f slot;
+            match Slot.Raw.next slot with
+            | None -> ()
+            | Some slot' -> iter_callstack_slot f slot'
+
+          (* Iterate over stack frames *)
+          let iter_callstack f bt =
+            for i = 0 to length bt - 1 do
+              iter_callstack_slot f (Slot.get bt i)
+            done
+        ]}
+    *)
+  end
+
+  val of_raw: Raw.t -> t
+  (** Extracts the user-friendly strack trace slot from a low-level raw slot. *)
+end with type stack_trace := t
+
+val slots: t -> Slot.t array option
+(** Returns the slots of a stack trace, or [None] if none of them contain useful
+    information.
```

Comment:
The presence of this option type seems a bit strange.  Maybe it would be better to always return the slots, and then have accessor functions that demonstrate what properties they have?

Reply:
OK, I threw away the option and and now return an empty array instead of `None`.